streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -271,27 +272,71 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
     }
 
     /**
-     * Get the receivers from the ReceiverInputDStreams, distributes them to the
-     * worker nodes as a parallel collection, and runs them.
+     * Get the list of executors excluding driver
      */
-    private def startReceivers() {
-      val receivers = receiverInputStreams.map(nis => {
-        val rcvr = nis.getReceiver()
-        rcvr.setReceiverId(nis.id)
-        rcvr
-      })
+    private def getExecutors(ssc: StreamingContext): List[String] = {
+      val executors = ssc.sparkContext.getExecutorMemoryStatus.map(_._1.split(":")(0)).toList
+      val driver = ssc.sparkContext.getConf.get("spark.driver.host")
+      executors.diff(List(driver))
+    }
 
-      // Right now, we only honor preferences if all receivers have them
+    /* Schedule receivers using preferredLocation if specified
+     * and round-robin otherwise
+     */
+    private def scheduleReceivers(receivers: Seq[Receiver[_]]): RDD[Receiver[_]] = {
+      // Location preferences are honored if all receivers have them
       val hasLocationPreferences = receivers.map(_.preferredLocation.isDefined).reduce(_ && _)
 
-      // Create the parallel collection of receivers to distributed them on the worker nodes
+      // If no location preferences are specified, set host location for each receiver
+      // so as to distribute them evenly over executors in a round-robin fashion
+      // If num_receivers > num_executors, distribute receivers among executors
+      // If num_executors > num_receivers, distribute executors among receivers
+      val locations = new Array[ArrayBuffer[String]](receivers.length)
+      var roundRobin = false
+      if (!hasLocationPreferences && !ssc.sparkContext.isLocal) {
+        val executors = getExecutors(ssc)
+        var i = 0
+        if (!executors.isEmpty) {
+          for (i <- 0 to (receivers.length - 1)) {
+            locations(i) = new ArrayBuffer[String]()
+          }
+          if (receivers.length >= executors.length) {
+            for (i <- 0 to (receivers.length - 1)) {
+              locations(i) += executors(i % executors.length)
+            }
+          } else {
+            for (i <- 0 to (executors.length - 1)) {
+              locations(i % receivers.length) += executors(i)
+            }
+          }
+          roundRobin = true
+        }
+      }
+
       val tempRDD =
         if (hasLocationPreferences) {
```

Comment:
As i said, in the other thread, these conditions become unnecessary if we come up with a single algorithm that works for all cases of receivers w/ or w/o locations.